### PR TITLE
Refactor install.py

### PR DESCRIPTION
**Pull request by sneekyfoxx**
- **Add** 'Mojo class' with the use of '__slots__'
- **Add** 'if__name__ == "\_\_main\_\_"'
- **Add** 'os.mkdir, os.symlink, os.chdir'
- **Use** 'shutil' for copy satisfying your TODO
- **Use** 'subprocess' when necessary

for other changes read the diff

### CODE REVIEW

**Comment by sneekyfoxx** [comment]
for the global install, *install.py* must be ran with **sudo** privileges because *shutil* with throw a **PermissionError**

**Comment by Sharktheone** [comment]
Thanks for the PR!
I like your changes. But I had not enough time to really review your changes, so I hope I can have a more in-depth look at your PR in the next few days.

File: PKGBUILD [hunk 1]
```diff
@@ -1,5 +1,5 @@
 pkgname=modular
-pkgver=0.4.1
+pkgver=0.7.0
```
Comment (Sharktheone) [inline PKGBUILD]: I think, you meant to update the Mojo version.

The newest modular version I have seen was 0.5.0, but they didn't announced it in any way (I think)

File: install.py [hunk 2]
```diff
@@ -5,243 +5,287 @@
...
+            stop = len(target)
+
+            while start < len(original):
+                if original[start:stop] == target:
+                    return True
+                start += 1
+                stop += 1
+
+            else:
+                return False
```
Comment (Sharktheone) [inline install.py]: What's the exact benefit of using this over pythons internal method with the `in` keyword? I assume, pythons own version is more efficient than a custom one?

File: install.py [hunk 3]
```diff
@@ -5,243 +5,287 @@
...
+                    sys.stdout.write("\nNo token provided")
+                    exit(1)
+
+                if self.token is not None:
+                    if self.token == "" or not self.token.startswith("mut_") or not len(self.token) == 36:
+                        sys.stdout.write("\nInvalid token")
+                        exit(1)
+
+                self.token = self.args[index]
+                self.skip_next_arg = True
```
Comment (Sharktheone) [inline install.py]: What I'm realizing now is that the `--dir` arg uses a `=` and here we aren't using the same syntax

File: install.py [hunk 4]
```diff
@@ -5,243 +5,287 @@
...
-            return None
-        elif yn == "other" or yn == "o":
-            return get_shell()
-        elif yn == "":
-            print("Skipping...")
+
+        try:
+            os.makedirs(self.working_dir)
+        except OSError:
+            pass
```
Comment (Sharktheone) [inline install.py]: I'm not sure if it's good to pass all errors here, it probably should only let `FileExistsError`s pass

File: install.py [hunk 5]
```diff
@@ -5,243 +5,287 @@
...
+            
+            if self.install_global:
+                shutil.copy(f"{self.working_dir}/lib/{self.arch}/libtinfo.so.6.4", "/usr/lib/")
+                os.symlink("/usr/lib/libtinfo.so.6.4", "/usr/lib/libtinfo.so.6")
+            else:
+                os.mkdir(f"{self.mojo_lib_path}")
+                shutil.copy(f"{self.working_dir}/lib/{self.arch}/libtinfo.so.6.4", f"{self.mojo_lib_path}/libtinfo.so.6")
+
+    def install_modular(self) -> None:
+        url = "https://raw.githubusercontent.com/sneekyfoxx/arch-mojo/seekyfoxx/PKGBUILD"
```
Comment (Sharktheone) [inline install.py]: I guess, it should be my repo here, so it's easier to update. Ah and it's actually the wrong PKGBUILD in the repo. I still need to delete this one. The right one is `aur/modular/PKGBUILD`

File: install.py [hunk 6]
```diff
@@ -5,243 +5,287 @@
...
+            os.symlink("/lib/libncurses.so.6.4", "/lib/libncurses.so.6")
+            os.symlink("/usr/lib/libform.so.6.4", "/usr/lib/libform.so.6")
+            os.symlink("/usr/lib/libpanel.so.6.4", "/usr/lib/libpanel.so.6")
+            os.symlink("/usr/lib/libedit.so.2.0.70", "/usr/lib/libedit.so.2")
+        else:
+            try:
+                if not os.path.exists(f"{self.mojo_lib_path}"):
+                    os.mkdir(f"{self.mojo_lib_path}")
+                else:
+                    os.makedirs(f"{self.mojo_lib_path}")
```
Comment (Sharktheone) [inline install.py]: I think, this code doesn't make sense

File: install.py [hunk 7]
```diff
@@ -5,243 +5,287 @@
...
+        if self.rc_pth is None:
+            sys.stdout.write("\nSkipping rc file installation")
+            self.print_manual_instructions()
+            exit(0)
 
-rc_pth = rc_path()
+        else:
+            with open(self.rc_pth, "a") as self.rc_file:
+                if self.rc_file is None:
+                    sys.stdout.write(f"C\nould not open {self.rc_pth}, skipping...")
```
Comment (Sharktheone) [inline install.py]: there's a typo here

**Review by Sharktheone — CHANGES_REQUESTED** [review]
I'm not sure, how you want to handle these changes. If you want, I can also fix some of them, but it's up to you to decide.

File: PKGBUILD (hunk 1 above)
Comment (sneekyfoxx) [reply PKGBUILD]: Yea, at first I wasn't sure, then I went back and changed it to 0.4.1 when I found that it was the modular version. I also see 0.5.0 as well in [this announcement](https://github.com/modularml/mojo/discussions/1214).

File: install.py (hunk 2 above)
Comment (sneekyfoxx) [reply install.py]: Yes, Python's `in` is faster than my custom `hasstr` and removing it would shorten the code as well.

File: install.py (hunk 5 above)
Comment (sneekyfoxx) [reply install.py]: Yea, I was testing something I should have undid the changes before I pushed it.

File: install.py (hunk 6 above)
Comment (sneekyfoxx) [reply install.py]: What needs to be changed here?

File: install.py (hunk 7 above)
Comment (sneekyfoxx) [reply install.py]: Okay, I see it

File: install.py (hunk 3 above)
Comment (sneekyfoxx) [reply install.py]: I kept your same syntax I just moved the function into the class.

File: install.py (hunk 4 above)
Comment (sneekyfoxx) [reply install.py]: Ok.

Comment (sneekyfoxx) [reply install.py]: Oh, I've actually changed it already. I'm going to push the current changes.

**Pull request closed without merging**